File: app/src/main/java/com/duncan/dpi/DeviceListActivity.java
package com.duncan.dpi;

import android.app.Activity;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.os.Build;
import android.os.Bundle;
import android.support.v7.app.ActionBarActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

import java.util.ArrayList;
import java.util.List;


public class DeviceListActivity extends ActionBarActivity {
    Toolbar toolbar;
    ListView list;
    List<Device> deviceList = new ArrayList<>();
    DeviceListAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_devicelist);

        //Vars
        toolbar = (Toolbar)findViewById(R.id.toolbar);
        list = (ListView)findViewById(R.id.deviceList);

        setSupportActionBar(toolbar);

        adapter = new DeviceListAdapter(this, deviceList);
        list.setAdapter(adapter);
        deviceList.add(new Device(Build.MANUFACTURER + " " + Build.MODEL + " (Your device)", CalcUtil.getWidth(this), CalcUtil.getHeight(this), CalcUtil.getScreenSize(this), true));
        deviceList.addAll(DeviceUtil.getDeviceList());
        adapter.notifyDataSetChanged();

        list.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                Intent resultIntent = new Intent();
                resultIntent.putExtra("device", deviceList.get(position));
                setResult(Activity.RESULT_OK, resultIntent);
                finish();
            }
        });
    }
}

File: app/src/main/java/com/duncan/dpi/activity/DeviceListActivity.kt
package com.duncan.dpi.activity

import android.annotation.TargetApi
import android.app.Activity
import android.content.Intent
import android.graphics.Point
import android.os.Build
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.duncan.dpi.adapter.DeviceRVAdapter
import com.duncan.dpi.databinding.ActivityDevicelistBinding
import com.duncan.dpi.`interface`.ItemClickListener
import com.duncan.dpi.model.Device
import com.duncan.dpi.util.CalcUtil
import com.duncan.dpi.util.DeviceUtil
import java.util.*


class DeviceListActivity : AppCompatActivity() {
    internal var deviceList: MutableList<Device> = ArrayList()
    private lateinit var binding: ActivityDevicelistBinding

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1) // getRealSize is actually available in API 14+. This warning sidesteps the bug where it's flagged otherwise
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityDevicelistBinding.inflate(layoutInflater)
        setContentView(binding.root)

        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val layoutManager =
            LinearLayoutManager(this@DeviceListActivity)
        binding.recyclerView.layoutManager = layoutManager
        binding.recyclerView.addItemDecoration(
            DividerItemDecoration(
                this@DeviceListActivity,
                layoutManager.orientation
            )
        )

        val point = Point()
        windowManager.defaultDisplay.getRealSize(point)
        val widthPixels = point.x
        val heightPixels = point.y
        val densityDPI = resources.displayMetrics.densityDpi

        deviceList.add(Device("${Build.MANUFACTURER} ${Build.MODEL} (Your device)",  widthPixels, heightPixels, CalcUtil.calculateScreenSize(widthPixels, heightPixels, densityDPI)))
        deviceList.addAll(DeviceUtil(this@DeviceListActivity).getDeviceList())

        val adapter = DeviceRVAdapter(deviceList)
        binding.recyclerView.adapter = adapter

        adapter.itemClickListener = object : ItemClickListener<Device> {
            override fun onItemClick(view: View, data: Device) {
                val resultIntent = Intent()
                resultIntent.putExtra("device", data)
                setResult(Activity.RESULT_OK, resultIntent)
                finish()
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when(item.itemId) {
            android.R.id.home -> {
                onBackPressed()
            }
        }
        return super.onOptionsItemSelected(item)
    }
}
